### ISSUE TITLE
Attatching sk to ggenie rom slot crashes.

### ISSUE THREAD

**Issue by reporter123** [mamedev/mame#10034]
The attachment slot of Sonic &  Knuckles is not supported by MAME if use in conjunction with Game Genie:

Ignoring MAME exception: Unknown slot option 'rom_sk' in slot 'mdslot:rom_ggenie:subslot'
Fatal error: Unknown slot option 'rom_sk' in slot 'mdslot:rom_ggenie:subKnuuslot'
Ignoring MAME exception: Unknown slot option 'rom_sk' in slot 'mdslot:rom_ggenie:subslot'
Unknown slot option 'rom_sk' in slot 'mdslot:rom_ggenie:subslot'
terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument

This does not match real system behavior which can handle this setup.

**Comment by FoxhackDN** [comment]
You're loading them backwards.

Cart 1 has to be sk; cart 2 has to be ggenie.

**Comment by reporter123** [comment]
>  If you load ggenie / ggenie1 first, then sk, it crashes. 

Aside from this what you see is exactly what the real hardware would do. Its the behavior I have here as well.

**Comment by FoxhackDN** [comment]
Tried it in 0.244, MAME just exits with `Unknown slot option 'rom_sk' in slot 'mdslot:rom_ggenie:subslot'` instead of crashing hard. So whatever is causing this crash was introduced in .245.

**Comment by reporter123** [comment]
The version have is .242 it just exits. If the console isn't there it likely tries  to print the error message but there's no visible result unless mame was launched from a command terminal. I should have been more careful with my wording.

**Comment by FoxhackDN** [comment]
MAMETesters report opened: https://mametesters.org/view.php?id=8369

**Comment by reporter123** [comment]
Just to clarify in a real system having sonic knuckles attached to a game genie cartilage is a valid configuration. As is then attaching a second cartilage to sonic knuckles. So it seems there's to issues going on here. With one being the crash for unsupported by emu  configurations. The other being that this configuration is not supported in the first place.

**Comment by FoxhackDN** [comment]
> The other being that this configuration is not supported in the first place.

Yes, it's not supported. The driver was updated to support several games that use non-standard save types; Sonic & Knuckles and Evander Holyfield's Real Deal Boxing were the two that I remembered, but there's like three more. The Game Genie handling code does not seem to have support for these special save types at the moment.

**Comment by angelosa** [comment]
Changed title, it's because ggenie_sub_cart function tries to do its own repeated version instead of just inheriting up `md_cart(device)` from md_carts.h like sk slot type does:

https://github.com/mamedev/mame/blob/4eb4dd2e87c5b8944c4308b2dc10cca01bef659c/src/devices/bus/megadrive/ggenie.cpp#L161

https://github.com/mamedev/mame/blob/4eb4dd2e87c5b8944c4308b2dc10cca01bef659c/src/devices/bus/megadrive/sk.cpp#L101

**Comment by cuavas** [comment]
The Mega Drive slot really needs a rewrite.  It’s too much of a leaky abstraction at the moment.  Probably not worth trying to work around this without going the whole hog.  The main catch is that there’s a line on the top slot (cartridge) that can reverse the chip selects for the top and bottom slots (cartridge and Mega CD).  Fortunately, I think it’s hardwired in every cartridge, so you don’t have to worry about it changing on-the-fly.
